mlflow/pyfunc/model.py

```diff
@@ -0,0 +1,379 @@
+                                                 mlflow_model=Model()):
+    """
+    Export model as a generic Python function model.
+    :param path: The path to which to save the Python model.
+    :param loader_module: The name of the Python module that will be used to load the model
+                          from ``data_path``. This module must define a method with the prototype
+                          ``_load_pyfunc(data_path)``.
+    :param data_path: Path to a file or directory containing model data.
+    :param code_paths: A list of local filesystem paths to Python file dependencies (or directories
+                      containing file dependencies). These files will be *prepended* to the system
+                      path before the model is loaded.
+    :param conda_env: Either a dictionary representation of a Conda environment or the path to a
+                      Conda environment yaml file. If provided, this decribes the environment
+                      this model should be run in. At minimum, it should specify the dependencies
+                      contained in :data:`mlflow.pyfunc.DEFAULT_CONDA_ENV`. If `None`, the default
+                      :data:`mlflow.pyfunc.DEFAULT_CONDA_ENV` environment will be added to the
+                      model.
+    :return: Model configuration containing model info.
+    """
+    if loader_module is None:
+        raise MlflowException(
+            message=("`loader_module` must be specified!"),
+            error_code=INVALID_PARAMETER_VALUE)
+
+    if os.path.exists(path):
+        raise MlflowException(
+                message="Path '{}' already exists".format(path),
+                error_code=RESOURCE_ALREADY_EXISTS)
+    os.makedirs(path)
+
+    code = None
+    data = None
+    env = None
+
+    if data_path is not None:
+        model_file = _copy_file_or_tree(src=data_path, dst=path, dst_dir="data")
+        data = model_file
+
+    if code_paths is not None:
+        for code_path in code_paths:
+            _copy_file_or_tree(src=code_path, dst=path, dst_dir="code")
+        code = "code"
+
+    if conda_env is not None:
+        shutil.copy(src=conda_env, dst=os.path.join(path, "mlflow_env.yml"))
+        env = "mlflow_env.yml"
+
+    mlflow.pyfunc.add_to_model(
+        mlflow_model, loader_module=loader_module, code=code, data=data, env=env)
+    mlflow_model.save(os.path.join(path, 'MLmodel'))
+    return mlflow_model
+
+
+def _validate_artifacts(artifacts):
+    """
+    Examines the specified artifacts and verifies that constituent models have compatible
+    versions of Python and compatible library dependency versions. If any incompatibilities are
+    detected, appropriate warnings are logged.
+
+    :param artifacts: A dictionary containing ``<name, artifact_path>`` entries, where
+                      ``artifact_path`` is the absolute filesystem path to the artifact.
+    """
+    try:
+        from conda.resolve import MatchSpec
```

Comment:
This seems relatively complicated to require a conda dependency (discoverable only at runtime) to do some conda-specific logic about resolving cloudpickle versions. What happens if cloudpickle is a transitive dependency?

Could we just print this warning at load time, as opposed to save time, like we do python version? Anyway, the runtime environment may not precisely match what we write (e.g., spark_udf does not currently use the conda env), so this is the only place we can be relatively sure, and anything else is just a complicated best guess.

Reply:
Whether or not we end up changing how we do this, I think it might be a good idea to move the validation logic and any associated tests to a follow-up PR, where we can discuss the complexities independently. Otherwise, it's a relatively minor aspect of this feature, but I think it contributes significantly to the complexity of the PR.

Also, a second point is that if the end goal of this logic is to print best-effort warnings, probably a failure to import conda should not be an exception, but just a warning that we could not do our validation.

Reply:
I agree that we shouldn't require a conda dependency for this validation logic. However, if the `conda.resolve` dependency happens to exist, I think it's reasonable to perform a check for conflicting library versions at model save time. If cloudpickle is a transitive dependency, it just means that we won't capture potentially conflicting state.

Checking for version compatibility at model load time also seems like a good idea. 

Keeping with your suggestion, I've removed the validation and associated tests from this PR; I'll file a separate one with some changes to the validation logic that avoids a strict dependency on `conda.resolve` and validates during model load.